notes/compilation-and-linkage.md

```diff
@@ -0,0 +1,684 @@
+# Compilation and Linkage
+
+Contributors: Tri Nguyen
+
+## Introduction
+
+Let's say I'm a structural engineer. I've made a drawing for a bridge, and I tell you to build it. How? Are you just gonna build it from left to right? Are you gonna build it part by part, and then assemble it?
+
+If you've used Eclipse/PyCharm/VSCode/etc., you make your program, press a button, and it magically runs. But how? In C/C++, it looks like this (prepare yourself):
+
+<div style="text-align: center">
+
+![](../images/g%2B%2B-compilation-process-1.PNG)
+
+Figure 1 - C/C++ Compilation
+
+</div>
+
+Just like how building a bridge will have a structured process, so will any language for compiling its program.
+
+**QUICK NOTE**: You may have noticed by now that C++ is compatible with C syntax, but not the other way around. The first 2 steps will be described with examples specific to C, but do note that C++ does the same thing, but it must use a different system (The C++ Native Compiler) which allows for the use of C++ specific-syntax.
+
+## The C Preprocessor
+
+For a bridge, let's say that one of the components is a beam. Me, the engineer, will need to draw out all the components of the beam. There might be things like screws, wires, and metal pieces that will be layed out in the drawing. 
+
+Likewise, as the coder, we must lay out all the necessary packages, functions, and definitions for our programs to run. The Preprocessor helps out with this a lot by simplifying the process.
+
+The C Preprocessor is a text-replacement program that modifies your code before it is converted into other forms. It contains registers/variables to keep track of certain things throughout the lifetime of its use.
+
+### Preprocessor Directives
+
+The preprocessor allows for the use of so-called "directives", which are various statements that do different things.
+
+#### Import (`#include`)
+
+This is quite a simple statement, which simply includes a file. To utilize the contents of a file, all you have to do is the following:
+
+<div style="text-align: center">
+
+`#include <libraryFile>`
+
+or
+
+`#include <libraryPath>`
+
+</div>
+
+You may have also created files that require access to definitions within other files you also created. In that case, you should use the following syntax
+
+<div style="text-align: center">
+
+`#include "myFile"`
+
+</div>
+
+Be wary that some libraries have files without `.c`, `.cpp`, `.h` or `.hpp`. Also note that you can never include an entire folder of C/C++ files. The reason being that whenever an `#include` directive occurs, it will replace it with the code in that file. That's right, no matter how big that file is, it will copy and paste it into your file, and folders don't directly contain code, so including a folder is not possible.
+
+#### Macros (`#define`, `#undef`)
+
+Macros are definitions used by the preprocessor, and these definitions are held within its registers. You can sort of think of them as variables, if that helps
```

Comment:
kinda pedantic, but what do you mean "these definitions are held within its registers"?

Also suggest writing "you can think of them as const variables"